apps/emqx_retainer/src/emqx_retainer_index.erl

```diff
@@ -0,0 +1,144 @@
+
+-spec select_index(emqx:words(), list(index())) -> index() | undefined.
+select_index(Tokens, Indexes) ->
+    select_index(Tokens, Indexes, 0, undefined).
+
+-spec condition(index(), emqx_topic:words()) -> ets:match_pattern().
+condition(Index, Tokens) ->
+    condition(Index, Tokens, 1, [], []).
+
+-spec condition(emqx_topic:words()) -> ets:match_pattern().
+condition(Tokens) ->
+    Tokens1 = [case W =:= '+' of true -> '_'; _ -> W end || W <- Tokens],
+    case length(Tokens1) > 0 andalso lists:last(Tokens1) =:= '#' of
+        false -> Tokens1;
+        _ -> (Tokens1 -- ['#']) ++ '_'
+    end.
+
+-spec restore_topic(index(), emqx_topic:words(), emqx_topic:words()) -> emqx_topic:words().
+restore_topic(Index, IndexTokens, OtherTokens) ->
+    restore_topic(Index, IndexTokens, OtherTokens, 1, []).
+
+%%--------------------------------------------------------------------
+%% Private
+%%--------------------------------------------------------------------
+
+split_index_tokens([NIndex | OtherIndex], [Token | Tokens], N, IndexTokens, OtherTokens)
+    when NIndex == N ->
+    split_index_tokens(OtherIndex, Tokens, N + 1, [Token | IndexTokens], OtherTokens);
+
+split_index_tokens([_NIndex | _] = Index, [Token | Tokens], N, IndexTokens, OtherTokens) ->
+    split_index_tokens(Index, Tokens, N + 1, IndexTokens, [Token | OtherTokens]);
+
+split_index_tokens([], Tokens, _N, IndexTokens, OtherTokens) ->
+    {lists:reverse(IndexTokens), lists:reverse(OtherTokens) ++ Tokens};
+
+split_index_tokens(_Index, [], _N, IndexTokens, OtherTokens) ->
+    {lists:reverse(IndexTokens), lists:reverse(OtherTokens)}.
+
+index_score([N | _Index], [Ph | _Tokens], N, Score)
+  when Ph =:= '+'; Ph =:= '#' -> Score;
+index_score([N | Index], [_Word | Tokens], N, Score) ->
+    index_score(Index, Tokens, N + 1, Score + 1);
+index_score(Index, [_Word | Tokens], N, Score) ->
+    index_score(Index, Tokens, N + 1, Score);
+index_score([], _Tokens, _N, Score) -> Score;
+index_score(_Index, [], _N, Score) -> Score.
+
+select_index(_Tokens, [], _MaxScore, SelectedIndex) -> SelectedIndex;
+select_index(Tokens, [Index | Indexes], MaxScore, SelectedIndex) ->
+    Score = index_score(Index, Tokens),
+    case Score > MaxScore of
+        true ->
+            select_index(Tokens, Indexes, Score, Index);
+        false ->
+            select_index(Tokens, Indexes, MaxScore, SelectedIndex)
+    end.
+
+condition([_NIndex | _OtherIndex], ['#' | _OtherTokens], _N, IndexMatch, OtherMatch) ->
+    {lists:reverse(IndexMatch) ++ '_', lists:reverse(OtherMatch) ++ '_'};
+
+condition([], ['#' | _OtherTokens], _N, IndexMatch, OtherMatch) ->
+    {lists:reverse(IndexMatch), lists:reverse(OtherMatch) ++ '_'};
+
+condition([], Tokens, _N, IndexMatch, OtherMatch) ->
+    {lists:reverse(IndexMatch), lists:reverse(OtherMatch) ++ condition(Tokens)};
+
+condition([_NIndex | _OtherIndex], [], _N, IndexMatch, OtherMatch) ->
+    {lists:reverse(IndexMatch) ++ '_', lists:reverse(OtherMatch)};
```

Comment:
`++ '_'` creates an improper list, is it intended?

Reply:
Yes, to match all possible tails.

Reply:
I see, thanks. Maybe add a comment?

Reply:
Yes, I think I should document all the `emqx_retainer_index` funs. I am to do this, but wanted to show the whole approach beforehand.